addons/stock/tests/test_move2.py

```diff
@@ -1810,6 +1810,73 @@ def test_empty_moves_validation_2(self):
         back_order = self.env['stock.picking'].search([('backorder_id', '=', delivery_order.id)])
         self.assertFalse(back_order, 'There should be no back order')
 
+    def test_pick_cancel_with_moves(self):
+
+        move = self.MoveObj.create({
+            'name': self.productA.name,
+            'product_id': self.productA.id,
+            'product_uom_qty': 10,
+            'product_uom': self.productA.uom_id.id,
+            'location_id': self.pack_location,
+            'location_dest_id': self.customer_location,
+        })
+
+        # create picking with stock move.
+        picking = self.env['stock.picking'].create({
+            'location_id': self.pack_location,
+            'location_dest_id': self.customer_location,
+            'partner_id': self.partner_delta_id,
+            'move_ids_without_package':move,
+            'picking_type_id': self.picking_type_out,
+            })
+
+        picking.action_confirm()
+        self.assertEqual(picking.state, 'confirmed', "Picking should be in a confirmed state.")
+
+        # picking in a confirmed state and try to cancel it.
+        picking.action_cancel()
+        self.assertEqual(picking.state, 'cancel', "Picking should be in a cancel state.")
+
+    def test_pick_cancel_without_moves(self):
+
+        # create picking wihtout stock move but with move_line and immediate transfer is 1.
+        picking = self.env['stock.picking'].create({
+            'location_id': self.pack_location,
+            'location_dest_id': self.customer_location,
+            'partner_id': self.partner_delta_id,
+            'picking_type_id': self.picking_type_out,
+            'immediate_transfer': True,
+            })
+
+        move_line = self.StockPackObj.create({
+            'product_id': self.productA.id,
+            'qty_done': 10,
+            'product_uom_id': self.productA.uom_id.id,
+            'location_id': self.pack_location,
+            'location_dest_id': self.customer_location,
+            'picking_id': picking.id,
+            })
+
+        location = self.env['stock.location'].browse(self.pack_location)
+
+        pack = self.env['stock.quant.package'].create({'name': 'The pack to pick'})
+        # make some stock
+        self.env['stock.quant']._update_available_quantity(self.productA, location, 10.0, package_id=pack)
+        quants = self.StockQuantObj.search([('product_id', '=', self.productA.id), ('location_id', '=', self.pack_location)])
+
+        package_level = self.env['stock.package_level'].create({
+            'package_id': pack.id,
+            'picking_id': picking.id,
+            'location_dest_id': self.customer_location,
+        })
+
+        picking.action_confirm()
+        self.assertEqual(picking.state, 'confirmed', "Picking should be in a confirmed state.")
+
+        # picking in a draft state and try to cancel it.
+        picking.action_cancel()
+        self.assertEqual(picking.state, 'cancel', "Picking should be in a cancel state.")
```

Comment:
what's the state of the stock move line after cancel? as it doesn't have any linked stock move..

i talked about with @amoyaux and he thinks we could maybe force create a cancel stock move the move lines without moves. the picking would directly be cancelled, the stock move line too and the reporting will be coreect

Reply:
@sle-odoo As we have stock move lines in immediate transfer ( always in the draft or done state: no confirm needed ) , If we generate movement based on the move lines it will increase stock moves ... which is not good.

it is just draft not confirmed, I don't know if it is useful for the reporting @lap-odoo

Reply:
@jir-odoo stock move are created when the picking is validated, why not when the picking is canceled ?

Reply:
@sle-odoo okay agree with reporting case, we can move with cancelling stock move not ( stock.move.line ), So the picking state will be 
auto change based on the move status.